aten/src/ATen/native/nested/NestedTensorTransformerFunctions.cpp

```diff
@@ -260,5 +261,128 @@ Tensor NestedTensor_to_mask(const Tensor& nt, c10::optional<int64_t> mask_dim, c
   }
   return result;
 }
+std::tuple<Tensor, int64_t> cumulative_and_max_seq_len(Tensor qkv) {
+  TORCH_CHECK(
+      qkv.is_nested(),
+      "QKV must be nested for flash cumulative_seq_len calculation.")
+  auto* nt_impl = get_nested_tensor_impl(qkv);
+  const auto& sizes = nt_impl->get_nested_size_tensor();
+  auto size_tensor_stride = sizes.stride(0);
+
+  const int64_t batch_size = qkv.size(0);
+  auto cumulative_seqlen = at::zeros({batch_size + 1}, TensorOptions().device(at::kCPU).dtype(at::kInt));
+
+  auto* sizes_ptr = sizes.data_ptr<int64_t>();
+  auto* cumulative_seqlen_ptr = cumulative_seqlen.data_ptr<int32_t>();
+
+  int32_t sum = 0;
+  int64_t max_seqlen = -1;
+  cumulative_seqlen_ptr[0] = sum;
+  for (const auto i : c10::irange(batch_size)) {
+    // Calculate the cumulative sum of the sequence lengths
+    auto current_seq_len = sizes_ptr[i * size_tensor_stride];
+    sum += current_seq_len;
+    cumulative_seqlen_ptr[i + 1] = sum;
+
+    // Find the max element while we traverse
+    max_seqlen = std::max(max_seqlen, current_seq_len);
+  }
+  // Send to GPU, this is pretty light weight calc for normal batch size
+  // but maybe this needs to be on gpu
+  cumulative_seqlen = cumulative_seqlen.to(TensorOptions().device(at::kCUDA));
+  return std::tuple<Tensor, int64_t>{cumulative_seqlen, max_seqlen};
+}
+
+struct {
+  Tensor q;
+  Tensor k;
+  Tensor v;
+  Tensor cumulative_sequence_length_q;
+  Tensor cumulative_sequence_length_k;
+  int64_t max_seqlen_q{};
+  int64_t max_seqlen_k{};
+} flash_attention_data;
+
+Tensor flash_attention_helper(
+    const Tensor& query,
+    const Tensor& key,
+    const Tensor& value,
+    int64_t num_heads,
+    double dropout_p,
+    bool causal) {
+  //  Query is of size (batch_size x ragged_seq_len x (potentially 3* n_heads *
+  //  head_dim))
+  auto cumulative_and_max_q = cumulative_and_max_seq_len(query);
+  Tensor cumulative_sequence_length_q = std::get<0>(cumulative_and_max_q);
+  int64_t max_seqlen_batch_q = std::get<1>(cumulative_and_max_q);
+
+  if (key.is_same(value) || query.is_same(key) || query.is_same(value)) {
+    int64_t head_dim{query.size(-1) / (num_heads * 3)};
+    int64_t Nnz_q{cumulative_sequence_length_q[-1].item<int64_t>()};
+
+    // The input/output of flash attention is (Nnz, nheads, headdim) we need to
+    // create a nested size with the last dim = nheads*headdim If q,k,v are all
+    // equal than the current nt last dim is of size nheads*headdim*3 after
+    // projection
+    auto atten_size = get_nested_size_tensor(query);
+    auto div3 = atten_size.index({at::indexing::Slice(), -1}).div(3);
+    atten_size.index({at::indexing::Slice(), -1}) = div3;
+
+    auto qkv_buffer_reshaped =
+        get_buffer(query).view({Nnz_q, 3, num_heads, head_dim});
```

Comment:
@jbschlosser 
The option to avoid using num_heads is 
` auto qkv_buffer_reshaped = get_buffer(query).view({Nnz_q, 3, 1, -1});`

The problem with this is that currently flash_attention has the restriction that:
`    TORCH_CHECK(head_size == 16 || head_size == 32 || head_size == 64 || head_size == 128, "Currently supported head_sizes: {16, 32, 64, 128}, passed in query head size of ", head_size);
`
So by doing this and avoiding num heads we just expand out the head_dim. Unfortunately for most reasonably sized q,k,v this expands the head dim passed 128.

I might be missing something: https://github.com/fairinternal/xformers/blob/cd8e3d218acc330029be5340fe6315df444b4bbf/xformers/ops.py#L339
This is what I was using as a reference.

Reply:
Sorry, I don't quite understand. I'd expect this for MHA (focusing on query but analogous for key / value):
1. Receive query of shape (N, L, E); N = original batch size, L = seq len, E = full embedding dim
2. Do query projection -> shape (N, L, E'), with E' = post-projection embedding dim
3. Reshape query to split heads -> shape (N, L, n_heads, head_dim); E' = n_heads * head_dim
4. Transpose query -> shape (N, n_heads, L, head_dim)
5. Combine query leading dims -> shape (N * n_heads, L, head_dim)
6. Pass query to SDP
7. Do reshaping + post-projection

So SDP should receive a query input of shape (N * n_heads, L, head_dim). Flash attention requires an input of (Nnz, n_heads, head_dim). IIUC xFormers sets n_heads=1 for passing to flash attention. So flash attention gets input of shape (N * n_heads * L, 1, head_dim). This should work fine with the head_dim constraints that flash attention has AFAICT.

Reply:
I changed assumptions on input which was a comment christian made.  This is an flash MHA nn module that does packed projection:
``` Python3
class FlashMHA(torch.nn.Module):
    def __init__(self, num_heads, qkv, proj):
        super().__init__()
        self.qkv = qkv
        self.proj = proj
        self.num_heads = num_heads

    def forward(self, query: torch.Tensor, key: torch.Tensor, value: torch.Tensor):
        assert query is key
        assert key is value

        query_projected = torch.nn.functional.linear(query, self.qkv.weight, self.qkv.bias)
        batch_size = query_projected.size(0)
        embed_dim = query_projected.size(-1)
        head_dim = embed_dim//(3*num_heads)
        
        query_projected = torch.reshape(query_projected, (batch_size, -1, 3, num_heads, head_dim))
        attn, _ = torch._scaled_dot_product_attention(query_projected, query_projected, query_projected)
        attn = attn.reshape((batch_size, -1, num_heads*head_dim))
        attn = torch.nn.functional.linear(attn, self.proj.weight, self.proj.bias)
        return attn, None
```

```
        #query_projected -> batch x (seq_len_q) x 3*num_heads*head_dim
        #query_projected -> batch x (seq_len_q) x 3 x num_heads x head_dim
        #attn -> batch x (seq_len_q) x 1 x num_heads x head_dim
        #attn -> batch x (seq_len_q) x num_heads*head_dim
  ```